kernel/smp.c

```diff
@@ -109,7 +109,7 @@ static void wait_for_start_signal(atomic_t *start_flag)
 static inline void smp_init_top(void *arg)
 {
 	struct k_thread dummy_thread;
-	struct cpu_start_cb *csc = arg;
+	struct cpu_start_cb csc = arg ? *(struct cpu_start_cb *)arg : (struct cpu_start_cb){0};
```

Comment:
This seems to me like it's only shrinking the window, not fixing the race, no?  You can still have two cores racing on the struct pointed to be arg, it just has to happen before this copy (where before it could happen any time before the pointer got followed below).

A correct fix would have to have the caller/spawner wait on something (semaphore, ad hoc spinlock, etc...) until the code here is able to signal "I'm finished and won't touch the arg struct anymore".

Reply:
@andyross no, don't think it's just shrinking the window, I think it's closing it. The thing is the `atomic_set()` which is signalling the primary core that we're up and that it can proceed. That's exactly the synchronisation mechanism that you're talking about, isn't it?

Reply:
Sorry, sick today so not as much time to review as I'd like.  Can you clarify this in the commit message then, something like "Make a copy of the object before <synchronization event> which guarantees <bad thing> can not happen"?

Reply:
Or alternatively explain the bug: "Previous code would have accessed <field> after <synchronization event> while another CPU might be <doing something racy>"

Reply:
I had a chance to come and look, and indeed, that's how it works.  Can we update the commit message to something like "The argument to smp_init_top() is a pointer to a static variable shared between different CPUs, it can't be used after wait_for_start_signal() returns, at which point it will hold the data for the LAST CPU started, not necessarily the one we were called with.  Make a copy."

Also: no need to test that pointer for null, this function and the call to smp_start_cpu() that launches it are both static to the same file, and the arg pointer is passed unconditionally.

Reply:
@andyross commit message updated, no changes otherwise